File: apps/ios/Sources/CrossyUI/ReactionStickerLayer.swift
// The reaction sticker layer as a VIEW overlay above the board Canvas (the owner's
// entry-shake finding, 2026-07-14). The first cut drew stickers inside the grid's
// per-frame Canvas pass, and the emoji visibly shivered through the entrance:
// Canvas re-renders its contents every TimelineView tick, so the glyph re-rasterized
// at every intermediate scale, and per-frame glyph rasterization re-snaps outline
// geometry sub-pixel (edge shimmer), where the web/CSS path rasterizes once and
// transforms the cached texture. This is the same lesson as the web's settle pop
// (#247), from the opposite direction: ANIMATE TRANSFORMS OF A RASTERIZED LAYER,
// NEVER RE-RENDER CONTENT PER FRAME. Here each sticker is one SwiftUI Text whose
// entrance, coalesce replay, and exit are scaleEffect/rotationEffect/opacity
// animations built from StickerEnvelope's constants: Core Animation transforms the
// glyph's one rasterized layer, the text itself never redraws mid-flight (its font
// size changes only with the camera's zoom, which re-renders the whole board
// anyway), and every keyframe track ends AT identity, so rest is exact by
// construction (the #245 guarantee). Moving off TimelineView also retires tick
// pacing as a shake contributor outright.
//
// Placement stays born-correct: position and tilt come from the sticker's seeded
// values and the camera alone, applied as un-animated layout, so nothing here can
// drift an incumbent. The layer is its own view over the grid: @Observable tracking
// keeps sticker mutations from re-evaluating CrossyGridView's body (only this body
// re-runs), and the layer never hit-tests, so every touch still belongs to the grid.

import Foundation
import SwiftUI

@available(iOS 17.0, macOS 14.0, *)
@MainActor
struct ReactionStickerLayer: View {
    let reactions: ReactionModel
    let puzzle: GridPuzzle
    let camera: GridCamera
    let reduceMotion: Bool

    var body: some View {
        ZStack {
            ForEach(reactions.stickers) { sticker in
                StickerView(
                    sticker: sticker,
                    fontSize: GridModule.stickerFontSize * camera.scale,
                    // A sticker that predates this layer by more than a beat (a
                    // resync of the lab's pre-settled twin, a placement swap) is
                    // already at rest: it mounts settled instead of shouting its
                    // entrance again, exactly as the closed form reads its age.
                    entersAnimated: Date().timeIntervalSinceReferenceDate - sticker.bornAt
                        < 0.1,
                    reduceMotion: reduceMotion
                )
                .position(position(of: sticker))
            }
        }
        .allowsHitTesting(false)
        .accessibilityHidden(true)
        // The sweep (the FlashBook pattern, scheduled): every mutation re-keys the
        // task, which sleeps to the book's soonest end and retires it; the ForEach
        // diff then removes the view, already faded by its own exit animation.
        .task(id: reactions.revision) {
            while !reactions.isEmpty, !Task.isCancelled {
                guard let expiry = reactions.nextExpiry else { return }
                let delay = expiry - Date().timeIntervalSinceReferenceDate
                if delay > 0 {
                    try? await Task.sleep(for: .seconds(delay + 0.02))
                }
                if Task.isCancelled { return }
                reactions.sweep(at: Date().timeIntervalSinceReferenceDate)
            }
        }
    }

    /// The sticker's anchor in view points: its seeded module-unit placement through
    /// the camera's transform. Un-animated layout, recomputed only when the camera
    /// or the book changes; during entrance/replay/exit this never moves.
    private func position(of sticker: ReactionSticker) -> CGPoint {
        let origin = GridModule.cellOrigin(sticker.cell, cols: puzzle.cols)
        return CGPoint(
            x: (origin.x + CGFloat(sticker.offsetX)) * camera.scale + camera.offset.x,
            y: (origin.y + CGFloat(sticker.offsetY)) * camera.scale + camera.offset.y)
    }
}

/// The web's per-segment easings as compositor curves, lifted from the same envelope
/// control points the Linux-tested closed forms evaluate: styles.css's --ease-out
/// token for the balloon and the settle, the CSS ease-in-out keyword for the tremble
/// legs and the dip. Non-isolated on purpose, like StickerEnvelope itself, so the
/// keyframes builder can read them from any context.
@available(iOS 17.0, macOS 14.0, *)
private enum LoudCurve {
    static let easeOut = curve(StickerEnvelope.entranceEaseOut)
    static let easeInOut = curve(StickerEnvelope.trembleEaseInOut)

    private static func curve(_ source: StickerEnvelope.EasingCurve) -> UnitCurve {
        .bezier(
            startControlPoint: UnitPoint(x: source.x1, y: source.y1),
            endControlPoint: UnitPoint(x: source.x2, y: source.y2))
    }
}

/// One sticker: a single Text whose animations are transforms of its rasterized
/// layer. Phases drive presence (born, resting, exiting) and one keyframe timeline
/// plays the loud gesture — the entrance AND every coalesce replay, a repeat shout
/// (owner ruling 2026-07-14) — over scale, tremble rotation, and opacity; the curves
/// are StickerEnvelope's constants, so the tested closed forms and the shipped
/// render share one source of truth.
@available(iOS 17.0, macOS 14.0, *)
@MainActor
private struct StickerView: View {
    let sticker: ReactionSticker
    let fontSize: CGFloat
    let entersAnimated: Bool
    let reduceMotion: Bool

    private enum Phase {
        case born
        case resting
        case exiting
    }

    /// The loud gesture's animated frame. Rest IS identity: the timeline's final
    /// keyframes land here exactly, so once a shout finishes the animator holds
    /// values indistinguishable from no modifier at all (nothing kept by a fill).
    private struct LoudFrame {
        var scale: Double = 1
        var trembleDegrees: Double = 0
        var opacity: Double = 1
    }

    @State private var phase: Phase
    /// Bumped once per shout: the entrance (from the task below, so a pre-settled
    /// mount that skips it never shouts) and every coalesce replay. The keyframe
    /// animator's trigger.
    @State private var shouts = 0

    init(sticker: ReactionSticker, fontSize: CGFloat, entersAnimated: Bool, reduceMotion: Bool) {
        self.sticker = sticker
        self.fontSize = fontSize
        self.entersAnimated = entersAnimated
        self.reduceMotion = reduceMotion
        // A settled arrival mounts at rest: no entrance replay (the lab's
        // settle-pair twin, or a sticker outliving a layer swap).
        _phase = State(initialValue: entersAnimated ? .born : .resting)
    }

    private var scale: Double {
        if reduceMotion { return 1 }
        switch phase {
        case .born, .resting: return 1
        case .exiting: return StickerEnvelope.exitFinalScale
        }
    }

    private var opacity: Double {
        switch phase {
        case .born: return 0
        case .resting: return 1
        case .exiting: return 0
        }
    }

    /// The animation for the CURRENT transition, evaluated when `phase` changes:
    /// into exiting it is the web's eased fade; under Reduce Motion the entrance
    /// (and a coalesce revival) is the plain fade. Otherwise nil ON PURPOSE — the
    /// loud keyframes own the whole entrance, so the born-to-resting hop must snap
    /// underneath them, never animate on its own.
    private var phaseAnimation: Animation? {
        if phase == .exiting {
            return .easeInOut(duration: StickerEnvelope.exitSeconds)
        }
        if reduceMotion {
            return .easeOut(duration: StickerEnvelope.reducedMotionFadeInSeconds)
        }
        return nil
    }

    var body: some View {
        Text(verbatim: sticker.emoji)
            .font(.system(size: fontSize))
            .rotationEffect(
                .degrees(StickerEnvelope.tiltDegrees(sticker, reduceMotion: reduceMotion)))
            // The loud gesture (owner ruling 2026-07-14, the web's sticker-in): fade
            // in small, balloon to 1.9x, tremble, settle — one keyframe timeline
            // replayed WHOLE on every shout. The tremble track rotates ON TOP of
            // the static tilt above (center rotations add, so the wobble swings
            // around the seeded lean, never replaces it), every track ends at
            // identity, and the opening MoveKeyframes jump each replay to the 0%
            // frame no matter where the last run ended.
            .keyframeAnimator(
                initialValue: LoudFrame(), trigger: shouts
            ) { [reduceMotion] content, frame in
                content
                    .scaleEffect(reduceMotion ? 1 : frame.scale)
                    .rotationEffect(.degrees(reduceMotion ? 0 : frame.trembleDegrees))
                    .opacity(reduceMotion ? 1 : frame.opacity)
            } keyframes: { _ in
                // Durations are spans between the envelope's absolute breakpoints
                // (the web rows at 16/26/38/50/60/68/100% of 1050 ms); tracks
                // shorter than the timeline hold their final value, which is the
                // resting one.
                KeyframeTrack(\LoudFrame.scale) {
                    MoveKeyframe(StickerEnvelope.entryFromScale)
                    LinearKeyframe(
                        StickerEnvelope.balloonScale,
                        duration: StickerEnvelope.balloonAt,
                        timingCurve: LoudCurve.easeOut)
                    LinearKeyframe(
                        StickerEnvelope.balloonScale,
                        duration: StickerEnvelope.settleDipStartAt - StickerEnvelope.balloonAt)
                    LinearKeyframe(
                        StickerEnvelope.settleDipScale,
                        duration: StickerEnvelope.settleDipAt - StickerEnvelope.settleDipStartAt,
                        timingCurve: LoudCurve.easeInOut)
                    LinearKeyframe(
                        1.0,
                        duration: StickerEnvelope.entranceSeconds - StickerEnvelope.settleDipAt,
                        timingCurve: LoudCurve.easeOut)
                }
                KeyframeTrack(\LoudFrame.trembleDegrees) {
                    MoveKeyframe(0)
                    LinearKeyframe(0, duration: StickerEnvelope.balloonAt)
                    LinearKeyframe(
                        StickerEnvelope.trembleBeats[0].degrees,
                        duration: StickerEnvelope.trembleBeats[0].at - StickerEnvelope.balloonAt,
                        timingCurve: LoudCurve.easeInOut)
                    LinearKeyframe(
                        StickerEnvelope.trembleBeats[1].degrees,
                        duration: StickerEnvelope.trembleBeats[1].at
                            - StickerEnvelope.trembleBeats[0].at,
                        timingCurve: LoudCurve.easeInOut)
                    LinearKeyframe(
                        StickerEnvelope.trembleBeats[2].degrees,
                        duration: StickerEnvelope.trembleBeats[2].at
                            - StickerEnvelope.trembleBeats[1].at,
                        timingCurve: LoudCurve.easeInOut)
                    LinearKeyframe(
                        StickerEnvelope.trembleBeats[3].degrees,
                        duration: StickerEnvelope.trembleBeats[3].at
                            - StickerEnvelope.trembleBeats[2].at,
                        timingCurve: LoudCurve.easeInOut)
                    LinearKeyframe(
                        StickerEnvelope.trembleBeats[4].degrees,
                        duration: StickerEnvelope.trembleBeats[4].at
                            - StickerEnvelope.trembleBeats[3].at,
                        timingCurve: LoudCurve.easeInOut)
                }
                KeyframeTrack(\LoudFrame.opacity) {
                    MoveKeyframe(0)
                    LinearKeyframe(
                        1.0,
                        duration: StickerEnvelope.balloonAt,
                        timingCurve: LoudCurve.easeOut)
                }
            }
            .scaleEffect(scale)
            .opacity(opacity)
            .animation(phaseAnimation, value: phase)
            // A coalesce replays the WHOLE loud gesture: a repeat shout, not a
            // softer echo. The trigger only fires on CHANGE, so birth itself bumps
            // `shouts` in the task below instead.
            .onChange(of: sticker.refreshedAt) { shouts += 1 }
            // The exit's clock: sleep to the sticker's final `exitSeconds`, then
            // fade out so removal (the model's sweep at endsAt) lands on an already
            // invisible view. Keyed on endsAt: a coalesce mid-life (or even
            // mid-exit) re-keys, revives the phase, and re-arms the exit against
            // the refreshed end. The born hop and the entrance shout land in one
            // transaction, so the first visible frame is already the 0% keyframe.
            .task(id: sticker.endsAt) {
                if phase == .born {
                    phase = .resting
                    shouts += 1
                }
                let exitStart = sticker.endsAt - StickerEnvelope.exitSeconds
                let delay = exitStart - Date().timeIntervalSinceReferenceDate
                if delay > 0 {
                    if phase == .exiting { phase = .resting }
                    try? await Task.sleep(for: .seconds(delay))
                }
                guard !Task.isCancelled else { return }
                phase = .exiting
            }
    }
}

File: apps/android/ui/src/main/kotlin/crossy/ui/ReactionStickerLayer.kt
// The reaction sticker layer as a Compose overlay ABOVE the board Canvas (the owner's entry-shake
// finding, 2026-07-14; twin of apps/ios ReactionStickerLayer and the web sticker layer). Each
// sticker is ONE Text of the native emoji glyph (owner ruling: platform font glyphs, no bundled
// images), and its entrance, coalesce replay, and exit are transforms of that rasterized layer via
// Modifier.graphicsLayer — the glyph is never re-rendered mid-flight. The transform values are
// sampled each frame from StickerEnvelope's closed forms, the SAME constants StickerEnvelopeTests
// pins, so the tested character and the shipped render share one source of truth (ANIMATE TRANSFORMS
// OF A RASTERIZED LAYER, NEVER RE-RENDER CONTENT PER FRAME).
//
// Placement is born-correct: each sticker's cell anchor and seeded tilt come from its own stable key
// alone, applied as un-animated layout, so a newcomer never drifts an incumbent. The layer never
// hit-tests (no pointerInput), so every touch still belongs to the grid beneath it.

package crossy.ui

import android.provider.Settings
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.offset
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.TransformOrigin
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.IntOffset
import androidx.compose.material3.Text
import androidx.compose.runtime.key
import androidx.compose.runtime.withFrameNanos
import androidx.compose.runtime.LaunchedEffect
import kotlin.math.roundToInt

/**
 * Draw every live sticker over the grid. `stickers` is the host's book (a pure `List` it mutates via
 * ReactionBook); `geometry` maps a cell index to its rect. The layer sizes to the grid: give it the
 * grid's own `fillMaxWidth().aspectRatio(cols/rows)` so its cell math matches CrossyGrid's exactly.
 *
 * `camera` is the grid's live transform (CrossyGrid.onCamera), threaded so the stickers ride the same
 * zoom and pan the board does (iOS passes its resolved GridCamera into the sticker layer). Null while
 * nothing has moved the board: at rest the layer keeps its own fit-to-width math, so nothing changes.
 * A non-null camera is already clamped by the grid, so its scale (dp per module unit) and offset (the
 * board origin in dp) place each sticker exactly where its cell renders, and the glyph grows with the
 * zoom just as the board's letters do.
 */
@Composable
fun ReactionStickerLayer(
    stickers: List<ReactionSticker>,
    geometry: GridGeometry,
    modifier: Modifier = Modifier,
    camera: GridCamera? = null,
    reduceMotion: Boolean = rememberReduceMotion(),
) {
    // Decorative: the ephemeral stickers are seen, not spoken (iOS ReactionStickerLayer
    // accessibilityHidden), so the emoji glyphs never reach the reader.
    BoxWithConstraints(modifier.clearAndSetSemantics {}) {
        val density = LocalDensity.current
        val cols = geometry.cols
        // At rest (no camera) the fit-to-width factor is the same one CrossyGrid derives when its own
        // camera is null: cell = width / cols, unit = cell / 36. With a camera the factor is the
        // camera's dp-per-unit scale converted to px, and the board origin shifts by the camera's dp
        // offset (both converted by display density). One `unitPx`/`originX`/`originY` covers both,
        // so the sticker anchor is `origin + moduleUnits * unitPx`, matching the grid's draw exactly.
        val d = density.density
        val restUnitPx = (with(density) { maxWidth.toPx() } / cols) / GridModule.UNIT
        val unitPx = if (camera != null) camera.scale * d else restUnitPx
        val originX = if (camera != null) camera.offsetX * d else 0f
        val originY = if (camera != null) camera.offsetY * d else 0f
        val fontSize = with(density) { (GridModule.STICKER_FONT_SIZE * unitPx).toSp() }

        for (sticker in stickers) {
            // A reaction anchored on a black square never renders (the web/iOS layer's rule).
            if (sticker.cell in geometry.blocks) continue
            val col = sticker.cell % cols
            val row = sticker.cell / cols
            val anchorX = originX + (col * GridModule.UNIT + sticker.offsetX.toFloat()) * unitPx
            val anchorY = originY + (row * GridModule.UNIT + sticker.offsetY.toFloat()) * unitPx
            key(sticker.id) {
                StickerGlyph(sticker, anchorX, anchorY, fontSize, reduceMotion)
            }
        }
    }
}

/**
 * One sticker: a single native-emoji Text whose scale, tremble rotation, and opacity are transforms
 * of its one rasterized layer, sampled each frame from StickerEnvelope's closed forms. The seeded
 * tilt is folded into the rotation the closed form returns (tremble composes OVER it). The glyph's
 * center sits at its cell anchor (in px, already through the camera); graphicsLayer recenters
 * (translate by -size/2) and transforms about that center, so nothing re-lays-out mid-flight.
 */
@Composable
private fun StickerGlyph(
    sticker: ReactionSticker,
    anchorX: Float,
    anchorY: Float,
    fontSize: androidx.compose.ui.unit.TextUnit,
    reduceMotion: Boolean,
) {
    // The per-frame sample of the shared monotonic clock (the same origin the book stamped bornAt
    // from). Reading it inside withFrameNanos ties the sample to the compositor's frame cadence
    // without re-rendering the glyph's content — only its transform changes.
    var now by remember(sticker.id) { mutableFloatStateOf(reactionNow().toFloat()) }
    LaunchedEffect(sticker.id) {
        while (true) {
            withFrameNanos { now = reactionNow().toFloat() }
        }
    }

    val t = now.toDouble()
    Text(
        text = sticker.emoji,
        style = TextStyle(fontFamily = FontFamily.Default),
        fontSize = fontSize,
        modifier = Modifier
            .offset { IntOffset(anchorX.roundToInt(), anchorY.roundToInt()) }
            .graphicsLayer {
                translationX = -size.width / 2f
                translationY = -size.height / 2f
                transformOrigin = TransformOrigin(0.5f, 0.5f)
                val s = StickerEnvelope.scale(sticker, t, reduceMotion).toFloat()
                scaleX = s
                scaleY = s
                rotationZ = StickerEnvelope.rotationDegrees(sticker, t, reduceMotion).toFloat()
                alpha = StickerEnvelope.opacity(sticker, t, reduceMotion).toFloat()
            },
    )
}

/** Whether the device has animations disabled (accessibility). Under Reduce Motion the sticker sits
 *  upright and fades only (owner spec), matching the web's prefers-reduced-motion fallback. Read
 *  from ANIMATOR_DURATION_SCALE == 0, Android's closest signal to prefers-reduced-motion. */
@Composable
fun rememberReduceMotion(): Boolean {
    val context = LocalContext.current
    return remember {
        Settings.Global.getFloat(
            context.contentResolver,
            Settings.Global.ANIMATOR_DURATION_SCALE,
            1f,
        ) == 0f
    }
}
